### PR TITLE
BLD: Check for reduce intrinsics and AVX512BW mask operations

### PR DESCRIPTION
**Pull request by seiko2plus**
MAINT::Distutils Test reduce intrinsics and AVX512BW mask operations

  Extending `Distutils::CompilerOpt` to allow adding extra separate test cases
  related to a certain CPU feature without affecting its availability.

  This patch also adds test cases for reduce intrinsics and AVX512BW mask operations,
  they can be reached through C #defentions  `NPY_HAVE_AVX512BW_MASK` and `NPY_HAVE_AVX512F_REDUCE`.

### CODE REVIEW

File: numpy/distutils/tests/test_ccompiler_opt_conf.py [hunk 1]
```diff
@@ -151,6 +152,24 @@ def test_group(self, error_msg, search_in, feature_name, feature_dict):
...
+        if not extra_checks:
+            return
+        if isinstance(extra_checks, str):
+            extra_checks = extra_checks.split()
+
+        for f in extra_checks:
+            impl_dict = search_in.get(f)
+            if not impl_dict or "disable" in impl_dict:
+                continue
+            raise AssertionError(error_msg + \
```
Comment (eric-wieser) [inline numpy/distutils/tests/test_ccompiler_opt_conf.py:168]: Backslash is not needed in parens
```suggestion
            raise AssertionError(error_msg +
```

File: numpy/distutils/ccompiler_opt.py [hunk 2]
```diff
@@ -243,7 +258,8 @@ class _Config:
         ),
         AVX512_SKX = dict(
             interest=42, implies="AVX512CD", group="AVX512VL AVX512BW AVX512DQ",
-            detect="AVX512_SKX", implies_detect=False
+            detect="AVX512_SKX", implies_detect=False,
+            extra_checks="AVX512BW_MASK"
```
Comment (eric-wieser) [inline numpy/distutils/ccompiler_opt.py:262]: Shouldn't this be 
```suggestion
            extra_checks=["AVX512BW_MASK"]
```

If not, where is the conversion from `str` to `list` occurring?

Comment (seiko2plus) [reply numpy/distutils/ccompiler_opt.py:262]: both acceptable `extra_checks=['x', 'xx']` or`extra_checks='x xx'`

> If not, where is the conversion from str to list occurring?

https://github.com/numpy/numpy/blob/97d2db483fc0ffd46f38d0e1c39d5fc001e33197/numpy/distutils/ccompiler_opt.py#L1139-L1144

Comment (eric-wieser) [reply numpy/distutils/ccompiler_opt.py:262]: Thanks, missed that because the line doing the splitting wasn't part of the diff

**Review by Qiyu8 — APPROVED** [review]

File: numpy/distutils/ccompiler_opt.py [hunk 3]
```diff
@@ -152,6 +152,18 @@ class _Config:
...
+            Extra test cases names for the CPU feature that need to be tested
+            against the compiler.
+
+            Each test case name must has a C file located at 'conf_check_path'
+            contains at least one intrinsic or function related to the test case
+            named extra_xxxx.c where xxxx is the case name in lower case.
+
+            If the compiler able to successfully compile the C file then `CCompilerOpt`
+            will add C #definition for it into the main dispatch header, e.g.
+            #define {conf_c_prefix}_XXXX where XXXX is the case name in upper case.
```
Comment (eric-wieser) [inline numpy/distutils/ccompiler_opt.py:165]: Grammar is slightly off here:
```suggestion
        "extra_checks": str or list, optional
            Extra test case names for the CPU feature that need to be tested
            against the compiler.

            Each test case must have a C file named ``extra_xxxx.c``, where
            ``xxxx`` is the case name in lower case, under 'conf_check_path'.
            It should contain at least one intrinsic or function related to the test case.

            If the compiler able to successfully compile the C file then `CCompilerOpt`
            will add C #definition for it into the main dispatch header, e.g.
            ```#define {conf_c_prefix}_XXXX`` where ``XXXX`` is the case name in upper case.
```

File: numpy/distutils/ccompiler_opt.py [hunk 4]
```diff
@@ -1487,6 +1503,45 @@ def feature_can_autovec(self, name):
...
+        self.dist_log("Testing extra checks for feature '%s'" % name, extra_checks)
+        flags = self.feature_flags(name)
+        available = []
+        not_available = []
+        for chk in extra_checks:
+            test_path = os.path.join(
+                self.conf_check_path, "extra_%s.c" % chk.lower()
+            )
+            if not os.path.exists(test_path):
+                self.dist_fatal("extra check file is not exist", test_path)
```
Comment (eric-wieser) [inline numpy/distutils/ccompiler_opt.py:1532]:
```suggestion
                self.dist_fatal("extra check file does not exist", test_path)
```

**Comment by Qiyu8** [comment]
@mattip This PR looks good to me, what do you think? some PR is stucked because of this one.

**Comment by eric-wieser** [comment]
I don't really have a good enough feel for this new distutils stuff to give a meaningful review beyond what I've already said and had addressed. Feel free to make the call @mattip.

**Comment by mattip** [comment]
Thanks @seiko2plus

**Pull request merged**